src/formatting_stack/util.clj

```diff
@@ -141,3 +143,34 @@
 
 (defn colorize [s color]
   (str \u001b (ansi-colors color) s \u001b (ansi-colors :reset)))
+
+(spec/def ::grouping-spec
+  (spec/cat ::neutral (spec/* (comp #{"NEUTRAL"} :lineType))
+            ::from    (spec/+ (comp #{"FROM"} :lineType))
+            ::rest    (spec/* any?)))
+
+(speced/defn ^coll? diff->line-numbers
+  "Returns maps of consecutive removals/changes (as `:begin` to `:end`) per `:filename` in `diff`."
+  [^string? diff]
+  (->> (io/input-stream (.getBytes diff))
+       (.parse (UnifiedDiffParser.))
+       (data/from-java)
+       (mapcat (fn [{:keys [toFileName hunks]}]
+                 (->> (loop [{::keys [from rest]}
```

Comment:
```clj
(loop [a
       x
       b y])
```

is inconsistent: better to keeps consistently horizontal or vertical, especially if there's a large `->>` in the middle